File: fixtures/scale/dart/large/lib/ciphers/lcg/lcg_cipher.dart
import 'dart:typed_data';

import '../../core/cipher.dart';
import '../../util/bytes.dart';
import '../../util/hex.dart';
import 'lcg_key.dart';

/// Keystream from a linear congruential generator seeded by the key.
class LcgCipher implements Cipher {
  final LcgKey key;

  LcgCipher(this.key);

  @override
  String name() => 'lcg';

  @override
  String encrypt(String plaintext) => hexEncode(_mask(bytesOf(plaintext)));

  @override
  String decrypt(String ciphertext) => bytesToText(_mask(hexDecode(ciphertext)));

  Uint8List _mask(Uint8List data) {
    var state = key.seed;
    final out = Uint8List(data.length);
    for (var i = 0; i < data.length; i++) {
      state = (state * 1103515245 + 12345) & 0xFFFFFFFF;
      out[i] = data[i] ^ ((state >> 16) & 0xFF);
    }
    return out;
  }
}

File: fixtures/scale/java/large/src/cryptobench/ciphers/lcg/LcgCipher.java
package cryptobench.ciphers.lcg;

import cryptobench.core.Cipher;
import cryptobench.util.Bytes;
import cryptobench.util.Hex;

/** Keystream from a linear congruential generator seeded by the key. */
public class LcgCipher implements Cipher {
    private final LcgKey key;

    public LcgCipher(LcgKey key) {
        this.key = key;
    }

    @Override
    public String name() {
        return "lcg";
    }

    @Override
    public String encrypt(String plaintext) {
        return Hex.encode(mask(Bytes.of(plaintext)));
    }

    @Override
    public String decrypt(String ciphertext) {
        return Bytes.toText(mask(Hex.decode(ciphertext)));
    }

    private byte[] mask(byte[] data) {
        long state = key.getSeed();
        byte[] out = new byte[data.length];
        for (int i = 0; i < data.length; i++) {
            state = state * 6364136223846793005L + 1442695040888963407L;
            out[i] = (byte) (data[i] ^ (state >>> 16));
        }
        return out;
    }
}
